File: experiments/repair-loop/testkits/001-hangman/python/solution.py
"""Phase-1.75 stub hangman solution for Python adapter validation.

Minimal working implementation: satisfies the 001-hangman API surface
with correct types and basic invariants, passes pyright (strict-mode off)
and the colocated harness tests. Not the canonical Phase-2/3 solution —
agents will write their own. Used here only to validate that the Python
target adapter wires pyright and pytest through the orchestrator without
schema-coupling defects.
"""
from __future__ import annotations

from dataclasses import dataclass, field, replace
from typing import Literal


GameStatus = Literal["playing", "won", "lost"]


@dataclass(frozen=True)
class GameState:
    secret: str
    guessed: frozenset[str] = field(default_factory=frozenset)
    remaining: int = 6
    status: GameStatus = "playing"


def initialize_game(secret: str) -> GameState:
    return GameState(secret=secret.lower())


def apply_guess(state: GameState, guess: str) -> GameState:
    if state.status != "playing":
        return state
    letter = guess.lower()
    if letter in state.guessed:
        return state
    new_guessed = state.guessed | {letter}
    if letter in state.secret:
        secret_letters = frozenset(state.secret)
        all_revealed = secret_letters <= new_guessed
        new_status: GameStatus = "won" if all_revealed else "playing"
        return replace(state, guessed=new_guessed, status=new_status)
    new_remaining = state.remaining - 1
    lost_status: GameStatus = "lost" if new_remaining <= 0 else "playing"
    return replace(
        state, guessed=new_guessed, remaining=new_remaining, status=lost_status
    )


def render_state(state: GameState) -> str:
    return " ".join(c if c in state.guessed else "_" for c in state.secret)


def game_status(state: GameState) -> GameStatus:
    return state.status

File: experiments/repair-loop/testkits/001-hangman/go/solution.go
// Phase-1.75 stub hangman solution for Go adapter validation.
//
// Module-mode upgrade: paired with go.mod and solution_test.go (both
// harness-owned via targets/go.json `harness_files`). Verifier chain
// runs `go vet ./...`, `go build ./...`, `go test ./...`. Not the
// canonical Phase-2/3 solution — agents will write their own.
//
// Naming convention: `Status` is the enum type (avoids collision with
// the `GameStatus` function); `GameStatus(state)` is the API accessor.
package main

import (
	"fmt"
	"strings"
)

type Status string

const (
	StatusPlaying Status = "playing"
	StatusWon     Status = "won"
	StatusLost    Status = "lost"
)

type GameState struct {
	Secret    string
	Guessed   map[rune]bool
	Remaining int
	Status    Status
}

func InitializeGame(secret string) *GameState {
	return &GameState{
		Secret:    strings.ToLower(secret),
		Guessed:   map[rune]bool{},
		Remaining: 6,
		Status:    StatusPlaying,
	}
}

func ApplyGuess(state *GameState, guess string) *GameState {
	if state.Status != StatusPlaying {
		return state
	}
	runes := []rune(strings.ToLower(guess))
	if len(runes) == 0 {
		return state
	}
	letter := runes[0]
	if state.Guessed[letter] {
		return state
	}
	newGuessed := make(map[rune]bool, len(state.Guessed)+1)
	for k, v := range state.Guessed {
		newGuessed[k] = v
	}
	newGuessed[letter] = true

	if strings.ContainsRune(state.Secret, letter) {
		allRevealed := true
		for _, c := range state.Secret {
			if !newGuessed[c] {
				allRevealed = false
				break
			}
		}
		status := StatusPlaying
		if allRevealed {
			status = StatusWon
		}
		return &GameState{
			Secret:    state.Secret,
			Guessed:   newGuessed,
			Remaining: state.Remaining,
			Status:    status,
		}
	}

	newRemaining := state.Remaining - 1
	status := StatusPlaying
	if newRemaining <= 0 {
		status = StatusLost
	}
	return &GameState{
		Secret:    state.Secret,
		Guessed:   newGuessed,
		Remaining: newRemaining,
		Status:    status,
	}
}

func RenderState(state *GameState) string {
	var b strings.Builder
	for i, c := range state.Secret {
		if i > 0 {
			b.WriteString(" ")
		}
		if state.Guessed[c] {
			b.WriteRune(c)
		} else {
			b.WriteString("_")
		}
	}
	return b.String()
}

func GameStatus(state *GameState) Status {
	return state.Status
}

func main() {
	// Smoke main: not exercised by the harness. Exists only so `go build`
	// in package-main mode produces an executable.
	state := InitializeGame("banana")
	state = ApplyGuess(state, "a")
	fmt.Println(RenderState(state))
}
